### PR TITLE
AI-16: Add ability to shutdown site when overdue

### PR DESCRIPTION
**Pull request by vaurdan**
## Description

Adds the ability to effectively shutdown a WordPress application, when the ~`WPCOM_VIP_OVERDUE_LOCKOUT`~ `VIP_OVERDUE_LOCKOUT` constant is set and truthy. 

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).

## Steps to Test

1. Check out PR.
2. Edit `config/wp-config.php` and add the following: 
```php
define( 'VIP_OVERDUE_LOCKOUT', true );
```
3. If testing on a VIP environment, you should see the `site-shutdown.html` page on all requests. 
4. If you are testing on a local environment, nothing should happen. Fake a VIP environment by adding 
```php
define('VIP_GO_ENV', true);
```
to your `wp-config.php` file

### CODE REVIEW

File: errors/site-shutdown.html
Comment (vaurdan) [inline errors/site-shutdown.html:145]: The copy here is temporary and will likely need a revision.

File: lib/sunrise/sunrise.php
Comment (vaurdan) [inline lib/sunrise/sunrise.php]: Should we use a different header here? I don't think there is any disadvantage on leveraging the same header, but let me know your thoughts.

File: 000-vip-init.php
Comment (rinatkhaziev) [inline 000-vip-init.php]: This can be simplified with `__return_true` rather than  using a closure

File: 000-vip-init.php
Comment (rinatkhaziev) [inline 000-vip-init.php]:
```suggestion
if ( Context::is_vip_env() && Context::is_overdue_locked() ) {
```

**Review by rinatkhaziev — DISMISSED** [review]
Left a couple of inline suggestions; aside from that, what do you think of dropping the `WPCOM_` part on the constants, the old ones are like that because of the legacy but probably we shouldn't keep on doing it.

File: lib/sunrise/sunrise.php
Comment (nickdaugherty) [reply lib/sunrise/sunrise.php]: I would say yes - the site isn't in maintenance mode so this header would be confusing / misleading.

File: lib/sunrise/sunrise.php
Comment (nickdaugherty) [inline lib/sunrise/sunrise.php]: 503 doesn't seem like the right status code, and will probably end up setting off various alarms and polluting metrics.

There's an "unofficial" 402 code for Payment Required that sounds right for this https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402

File: 000-vip-init.php
Comment (mjangda) [inline 000-vip-init.php:15]: Can we add a comment noting why we're doing this?

File: 000-vip-init.php
Comment (mjangda) [inline 000-vip-init.php]:
```suggestion
			header( 'X-VIP-402: true' );
```

File: errors/site-shutdown.html
Comment (mjangda) [inline errors/site-shutdown.html]:
```suggestion
					<span class="error">402</span>
```

File: errors/site-shutdown.html
Comment (mjangda) [inline errors/site-shutdown.html]:
```suggestion
				<p>This site is currently not available. Please try again later.</p>
```

File: errors/site-shutdown.html
Comment (mjangda) [inline errors/site-shutdown.html]:
```suggestion
			_error = 'vip-site-402';
```

File: lib/sunrise/sunrise.php
Comment (mjangda) [reply lib/sunrise/sunrise.php]: Will 402s be cached in Varnish? Will need to look into how we handle that (e.g. if cached, we probably want to flush the cache after the lockout is removed)

File: lib/sunrise/sunrise.php
Comment (mjangda) [reply lib/sunrise/sunrise.php]:
```suggestion
			header( 'X-VIP-402: true' );
```

File: errors/site-shutdown.html
Comment (mjangda) [inline errors/site-shutdown.html]:
```suggestion
		<title>Site Unavailable</title>
```

**Review by mjangda — DISMISSED** [review]
Added some suggestions for the language / wording.

File: errors/site-shutdown.html
Comment (nickdaugherty) [reply errors/site-shutdown.html]: Good question about 402 cacheability. I don't see any special rules for 4xx, aside from 404 which is cached for 1 minute, so I think our normal cache logic will be in play.

File: lib/sunrise/sunrise.php
Comment (nickdaugherty) [reply lib/sunrise/sunrise.php]: Replied in another spot, but I couldn't find any exceptions in varnish.vcl that would affect 402, so AFAIK it will receive the normal cache rules, which means we will need to do a cache flush on lockout...and on lockout for that matter.

Comment (vaurdan) [reply lib/sunrise/sunrise.php]: That's a very good point, and thanks for checking @nickdaugherty! I'm going to add the extra step for the cache flush on the respective GOOP endpoints.

File: 000-vip-init.php
Comment (vaurdan) [reply 000-vip-init.php:15]: @sjinks I believe you added this line. I would say it is related with the ~SonarCloud~CodeCov report, but I wanted to double check and ask for your input before adding a comment here.

Comment (sjinks) [reply 000-vip-init.php:15]: @mjangda please disregard this for now - I will write a P2 post explaining all these things.

In brief, PHPUnit is unable to generate a coverage report for the files loaded from the bootstrap script, even if those files are used in tests (I suspect that the code needs to be instrumented when it is compiled, and compilation of the bootstrap script and the files it requires happens before the code coverage driver is initialized). This is why these files are reported as uncovered.

I want to try another approach: initialize the autoloader and load the test library in `bootstrap.php` and use PHPUnit's Test Listener feature to load all mu-plugins before the very first test (the driver should be initialized by that time). However, I am not sure this will play nice with the way the WordPress Test Library initializes all hooks and filters.

**Comment by vaurdan** [comment]
Is there anything else that needs to be addressed before we can merge this PR to `develop`?

File: tests/lib/utils/context.php
Comment (nickdaugherty) [inline tests/lib/utils/context.php]:
```suggestion
		$this->assertFalse( $actual_result, 'The `VIP_OVERDUE_LOCKOUT` constant is not defined, but site is reported as locked' );
```

The text of the error message doesn't reflect what's being asserted.

File: tests/lib/utils/context.php
Comment (nickdaugherty) [inline tests/lib/utils/context.php]: If we want to keep with the theme of having a custom error message for the assertion, we can do this:

```suggestion
		$this->assertTrue( $actual_result, 'The `VIP_OVERDUE_LOCKOUT` constant is set to `true`, but site is not reported as locked' );
```

File: lib/utils/class-context.php
Comment (nickdaugherty) [inline lib/utils/class-context.php:27]: For simplicity, we can skip the `constant()` function here, as it's only really needed when the constant name is not known beforehand (a variable).
```suggestion
			&& true === VIP_OVERDUE_LOCKOUT;
```

File: lib/sunrise/sunrise.php
Comment (nickdaugherty) [inline lib/sunrise/sunrise.php]: Looking at this closer, I wonder if we actually need this code block at all.

The regular locked / not locked check happens separately in 000-vip-init.php ... do we need to handle it again here too? The cross section of sites that are 1) locked out, and 2) have site not found errors is probably pretty tiny, and worst case, the end user is still seeing an error page and not real content.

Just trying to reduce the amount of code and duplication we have, if it makes sense.

File: 000-vip-init.php
Comment (nickdaugherty) [inline 000-vip-init.php]: Do we need any exceptions for logged-in VIP members? Could be a future enhancement I suppose. Just thinking of the case where someone on our end needs to do _something_ on a site while it's in lockout.

Folks that have the ability to sandbox sites can hack around it, just wondering if people think it's worthwhile to account for now.

**Comment by nickdaugherty** [comment]
@vaurdan I left a few more non-blocking thoughts, I think this is really close.

File: lib/utils/class-context.php
Comment (rinatkhaziev) [reply lib/utils/class-context.php:27]: @nickdaugherty, we prefer `constant()` these days because it's better testable.

File: lib/sunrise/sunrise.php
Comment (vaurdan) [reply lib/sunrise/sunrise.php]: That's a very good point. I think that I will remove this from the sunrise.php. As you said, worse case scenario, an error will still be displayed instead of any content.

File: 000-vip-init.php
Comment (vaurdan) [reply 000-vip-init.php]: Good point! I have created AI-22 to track this improvement.

File: lib/sunrise/sunrise.php
Comment (nickdaugherty) [inline lib/sunrise/sunrise.php]:
```suggestion
```

@vaurdan we have a stray line, now that we've removed the other code.

**Review by nickdaugherty — DISMISSED** [review]
This looks great!

File: lib/sunrise/sunrise.php
Comment (vaurdan) [reply lib/sunrise/sunrise.php]: Good catch! :D

**Review by aswasif007 — DISMISSED** [review]

**Comment by aswasif007** [comment]
@vaurdan Let's merge it and try locking out a test site.

**Comment by vaurdan** [comment]
Just waiting for the final approval by @Automattic/vip-platform-cantina :)

**Comment by nickdaugherty** [comment]
Tested and works great! @Automattic/vip-platform-cantina any help you can offer to get this across the finish line is greatly appreciated 🙇

**Review by WPprodigy — DISMISSED** [review]
Looks good! Sorry for the delayed review here.

Couple (non-blocking) thoughts:

1) Lets double check that it is cacheable, and ensure the cache is purged when a site is unlocked.

2) Mozilla says "This response code is reserved for future use": https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses. Seems adoption for it was too low. But also, super unlikely there is any movement on this, as these things seem to rarely change 😆 

3) Disabling cron means we won't be getting SDS syncs. Having that data fall out of sync isn't super ideal, but atm there's not a means to only run one type of job. (could be a fun future project to have a backup means of syncing to SDS for when cron is turned off / broke)

**Comment by nickdaugherty** [comment]
Thanks for looking @WPprodigy !

> Lets double check that it is cacheable, and ensure the cache is purged when a site is unlocked.

I've looked at Varnish VCL and 4xx is cacheable AFAICT ... the only specific handling we have is for 404, which is cached for 60s, otherwise it should behave the same.

> But also, super unlikely there is any movement on this, as these things seem to rarely change

It's easy enough to change in the future if needed, but this code makes the most sense IMO since it's 1) in the 4xx error range, and 2) related to payments (a subtle jab 😄) 

> Disabling cron means we won't be getting SDS syncs.

Very good point - I think we're safe to leave cron running...it's not really the target of these lockouts. @vaurdan what do you think about leaving cron alone?

**Comment by vaurdan** [comment]
> Disabling cron means we won't be getting SDS syncs. Having that data fall out of sync isn't super ideal, but atm there's not a means to only run one type of job. (could be a fun future project to have a backup means of syncing to SDS for when cron is turned off / broke)

Great point! I just committed a change to ignore the lockout when running from WP-CLI, cron included.

**Review by WPprodigy — APPROVED** [review]

**Pull request merged**
